hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/CustomFeeAssessmentStep.java

```diff
@@ -16,14 +16,267 @@
+
+        if (!result2.haveAssessedChanges()) {
+            recordBuilder.assessedCustomFees(customFeesAssessed);
+            return List.of(modifiedInputBody, level1Body);
+        }
+
+        /* -------------- Level 3 - Since there can only be 3 levels , no assessment here -------- */
+
+        // increment the level to create transaction body from all custom fees assessed from original
+        // transaction
+        validateTrue(++levelNum <= maxCustomFeeDepth, CUSTOM_FEE_CHARGING_EXCEEDED_MAX_RECURSION_DEPTH);
+        final var level2Body = buildBodyFromAdjustments(result2);
+        validateTotalAdjustments(level2Body, maxTransfersDepth);
+
+        recordBuilder.assessedCustomFees(customFeesAssessed);
+        return List.of(modifiedInputBody, level1Body, level2Body);
+    }
+
+    private AssessmentResult setUpNextLevelResult(
+            final CryptoTransferTransactionBody prevLevelBody, final AssessmentResult result) {
+        final var nextLevelResult = new AssessmentResult(
+                prevLevelBody.tokenTransfersOrElse(emptyList()),
+                prevLevelBody.transfersOrElse(TransferList.DEFAULT).accountAmountsOrElse(emptyList()));
+        nextLevelResult.setExemptDebits(result.getExemptDebits());
+        nextLevelResult.setRoyaltiesPaid(result.getRoyaltiesPaid());
+        return nextLevelResult;
+    }
+
+    private void validateTotalAdjustments(final CryptoTransferTransactionBody op, final int maxTransfersDepth) {
+        final var hbarTransfers = op.transfersOrElse(TransferList.DEFAULT)
+                .accountAmountsOrElse(emptyList())
+                .size();
+        var fungibleTokenChanges = 0;
+        var nftTransfers = 0;
+        for (final var xfer : op.tokenTransfersOrElse(emptyList())) {
+            fungibleTokenChanges += xfer.transfersOrElse(emptyList()).size();
+            nftTransfers += xfer.nftTransfersOrElse(emptyList()).size();
+        }
+
+        totalBalanceChanges += hbarTransfers + fungibleTokenChanges + nftTransfers;
+        validateTrue(totalBalanceChanges <= maxTransfersDepth, CUSTOM_FEE_CHARGING_EXCEEDED_MAX_ACCOUNT_AMOUNTS);
+    }
+
+    private CryptoTransferTransactionBody changedInputTxn(
+            final CryptoTransferTransactionBody op, final AssessmentResult result) {
+        final var copy = op.copyBuilder();
+        final var changedTokenTransfers = result.getInputTokenAdjustments();
+        final List<AccountAmount> aaList = new ArrayList<>();
+        final List<TokenTransferList> tokenTransferLists = new ArrayList<>();
+        // If there are no changes for the token , add as it is
+        for (final var xfers : op.tokenTransfers()) {
+            final var token = xfers.token();
+            if (!changedTokenTransfers.containsKey(token)) {
+                tokenTransferLists.add(xfers);
+            }
+        }
+        // If there are changes modify the token transfer list
+        for (final var entry : changedTokenTransfers.entrySet()) {
```

Comment:
Doesn't `changedTokenTransfers` also include non-fungible token transfers? I'm only seeing account amount (i.e. fungibles) added to the token transfer list in this `for` loop